### PR TITLE
Make publishing using AzDO release pipelines an opt-in feature.

### PR DESCRIPTION
**Pull request by JohnTortugo**
Related to these: https://github.com/dotnet/arcade/issues/1545 https://github.com/dotnet/arcade/issues/2084

- Goal here is to add a new parameter to Publish.proj so that the user can opt into using AzDO release pipelines for publishing. 
- When using AzDO pipelines for publishing, PublishPackagesToBlobFeed and PublishFilesToBlobFeed will be executed by the pipeline itself.
- Also added a new parameter to Build.proj to receive the opt in parameter.
- For now, my plan is that we can use a AzDO queue time variable (PUBLISH_USING_PIPELINES) to quick test the feature in any repo. Eventually, repos that want to opt into using the feature can just hard code the value in the build definition.

### CODE REVIEW

File: src/Microsoft.DotNet.Arcade.Sdk/tools/Build.proj [hunk 1]
```diff
@@ -80,6 +81,9 @@
     <_PublishToBlobStorage>false</_PublishToBlobStorage>
     <_PublishToBlobStorage Condition="'$(DotNetPublishToBlobFeed)' == 'true'">true</_PublishToBlobStorage>
```
Comment (riarenas) [inline src/Microsoft.DotNet.Arcade.Sdk/tools/Build.proj]: Should this be false if PUBLISH_USING_PIPELINES is set?

File: src/Microsoft.DotNet.Arcade.Sdk/tools/Build.proj [hunk 2]
```diff
@@ -80,6 +81,9 @@
...
 
+    <_PublishUsingPipelines>false</_PublishUsingPipelines>
```
Comment (chcosta) [inline src/Microsoft.DotNet.Arcade.Sdk/tools/Build.proj]: Does this have to be inited to 'false'? or can you just delete this line?  For that matter, can you get rid of `_PublishUsingPipelines` and `_PublishToblobStorage` and use the `DotNetPublishToBlobFeed` and `DotNetPublishUsingPipelines` directly instead?

File: src/Microsoft.DotNet.Arcade.Sdk/tools/Build.proj (hunk 1 above)
Comment (JohnTortugo) [reply src/Microsoft.DotNet.Arcade.Sdk/tools/Build.proj]: For the publishing it doesn't matter actually. Also, this `_PublishToBlobStorage ` is used during restoring to know if the Tasks.Feed package needs to be installed. So I'm thinking to leave this for now.

**Comment by chcosta** [comment]
Are there other things in Arcade that follow this model?

> For now, my plan is that we can use a AzDO queue time variable (PUBLISH_USING_PIPELINES) to quick test the feature in any repo. Eventually, repos that want to opt into using the feature can just hard code the value in the build definition.

**Comment by JohnTortugo** [comment]
> Are there other things in Arcade that follow this model?
> 
> > For now, my plan is that we can use a AzDO queue time variable (PUBLISH_USING_PIPELINES) to quick test the feature in any repo. Eventually, repos that want to opt into using the feature can just hard code the value in the build definition.

I don't think so. Please let me know if you have a suggestion about how to improve this.

**Comment by riarenas** [comment]
> > Are there other things in Arcade that follow this model?
> > > For now, my plan is that we can use a AzDO queue time variable (PUBLISH_USING_PIPELINES) to quick test the feature in any repo. Eventually, repos that want to opt into using the feature can just hard code the value in the build definition.
> 
> I don't think so. Please let me know if you have a suggestion about how to improve this.

Have we had any other cases where we wanted to do something similar? We're basically trying to create a 'feature flag' that we can phase out once we are ready to move all publishing to the pipelines model.
A build time variable seems like a good candidate for this as it is lightweight. An alternative could be adding a parameter to the yaml templates, but it seems a little heavy handed for something that will be disabled afterwards.

**Comment by chcosta** [comment]
We used to have the `PublishType` concept where you could say `PublishType=[blob]` or `PublishType=[blob,pipeline]`.  I'm not sure if that's better or not.  I think, the thing that smacks me in the face here is that "PUBLISH_USING_PIPELINES" doesn't follow the typical convention for variables naming.

**Long term, is the intention to enable this by default for builds running in Azure DevOps, or to always be opt-in?**

What's the downside of just always publishing artifacts if "SYSTEM_TEAMFOUNDATIONCOLLECTIONURI" is defined?  It seems like you could get rid of the `PublishToAzureDevOpsArtifacts` concept, remove the conditions from Publish.proj

```XML
    <PublishDependsOnTargets Condition="$(PublishToAzure) and '$(PublishToBlobFeedFlatContainer)' != 'true'">$(PublishDependsOnTargets);PublishPackagesToBlobFeed</PublishDependsOnTargets>
    <PublishDependsOnTargets Condition="$(PublishToAzure) and '$(PublishToBlobFeedFlatContainer)' == 'true'">$(PublishDependsOnTargets);PublishFilesToBlobFeed</PublishDependsOnTargets>
    <PublishDependsOnTargets Condition="$(PublishToSymbolServer)">$(PublishDependsOnTargets);PublishSymbols</PublishDependsOnTargets>
    <PublishDependsOnTargets Condition="$(PublishToSourceBuildStorage)">$(PublishDependsOnTargets);PublishToSourceBuildStorage</PublishDependsOnTargets>
    <PublishDependsOnTargets Condition="$(PublishToAzure)">$(PublishDependsOnTargets);PublishToAzureDevOpsArtifacts</PublishDependsOnTargets>
    <PublishDependsOnTargets>BeforePublish;$(PublishDependsOnTargets)</PublishDependsOnTargets>
```

...and update the target condition

```XML
<Target Name="PublishToAzureDevOpsArtifacts"
              Condition="'$(SYSTEM_TEAMFOUNDATIONCOLLECTIONURI)' != '' and '$(_EnablePublishToAzureDevOpsArtifacts)' == 'true'">
```

Then azure-pipelines.xml looks like


```YAML
        - _InternalBuildArgs: /p:DotNetSignType=$(_SignType) /p:TeamName=$(_TeamName)
            /p:DotNetPublishBlobFeedKey=$(dotnetfeed-storage-access-key-1)
            /p:DotNetPublishBlobFeedUrl=$(_PublishBlobFeedUrl)
            /p:DotNetPublishToBlobFeed=$(_DotNetPublishToBlobFeed)
            /p:OfficialBuildId=$(BUILD.BUILDNUMBER)

            ...

      steps:
      - checkout: self
        clean: true
      # Use utility script to run script command dependent on agent OS.
      - script: $(_Script)
          -configuration $(_BuildConfig) 
          -prepareMachine
          $(_InternalBuildArgs)
          $(_ValidateSdkArgs)
          /p:Test=false
        displayName: Windows Build / Publish
        env:
          _EnablePublishToAzureDevOpsArtifacts: true
```

And at some point, we just remove the `_EnablePublishToAzureDevOpsArtifacts` from the condition on the `PublishToAzureDevOpsArtifacts` target so that it's enabled by default when running in Azure DevOps.

If this is always going to be opt-in, then yeah, we probably need to maintain the `/p:DotNetPublishUsingPipelines` concept.

**Comment by riarenas** [comment]
> Long term, is the intention to enable this by default for builds running in Azure DevOps, or to always be opt-in?

The intention is for the opt-in to be short lived while we validate that the async publishing model is working correctly.

> What's the downside of just always publishing artifacts if "SYSTEM_TEAMFOUNDATIONCOLLECTIONURI" is defined?

If we publish the artifacts to azure devops for every build, we might get in a state where we try to publish twice. Once during the build, and then again during a release pipeline. That might be fine while we're just trying to enable things though.

**Comment by JohnTortugo** [comment]
> **Long term, is the intention to enable this by default for builds running in Azure DevOps, or to always be opt-in?**

Short term it will be opt-in. If things goes smooth it will become the default for publishing. Once it become the default I think we won't need any condition for checking if we should run it or not as it will run whenever publishing should be run. Note that publishing of other assets (e.g., symbols) will also be moved to pipelines eventually.

> the thing that smacks me in the face here is that "PUBLISH_USING_PIPELINES" doesn't follow the typical convention for variables naming.

I can easily change that. I'm not sure what's the convention for definition variables. Would _PublishUsingPipelines be better?

**Comment by chcosta** [comment]
> I can easily change that. I'm not sure what's the convention for definition variables.

Does that mean you'll go with this?

```YAML
      steps:
      - checkout: self
        clean: true
      # Use utility script to run script command dependent on agent OS.
      - script: $(_Script)
          -configuration $(_BuildConfig) 
          -prepareMachine
          $(_InternalBuildArgs)
          $(_ValidateSdkArgs)
          /p:Test=false
        displayName: Windows Build / Publish
        env:
          _PublishUsingPipelines: true
```

**Comment by JohnTortugo** [comment]
> Does that mean you'll go with this?

I think so, but we want it false by default. I assume that this way we'll still be able to override `_PublishUsingPipelines` at queue time, right? I've opted to make the activate/deactivate switch a queue time variable to let the user quickly switch - i.e., no update/PR to .yml would be needed.

**Comment by chcosta** [comment]
If you want this defined at queue time, You should probably just omit the "env" block from the yaml and add a variable to the build definition.  variables get confusing when you're defining them in yaml and in the build definition.  In the past, yaml variables took precedence over build definition variables.  I'm not sure if that has changed.  The model I proposed above would mostly work along those lines and you wouldn't have any yaml changes.

**Comment by JohnTortugo** [comment]
> If you want this defined at queue time, You should probably just omit the "env" block from the yaml and add a variable to the build definition.

Pushed new change where I rename the variable and don't use an `env` block to define it. Also, I want to point out that I already tested the SDK/YML produced from this PR on this build https://dnceng.visualstudio.com/internal/_build/results?buildId=115429 where I defined the PUBLISH_USING_PIPELINES variable and the switch worked fine.

**Comment by chcosta** [comment]
> Also, I want to point out that I already tested the SDK/YML produced from this PR on this build https://dnceng.visualstudio.com/internal/_build/results?buildId=115429 where I defined the PUBLISH_USING_PIPELINES variable and the switch worked fine.

Yes, that's what I was saying would work except I didn't think you needed to explicitly pass that variable in yaml or define it in a variables section of yaml (you weren't doing the latter).

File: src/Microsoft.DotNet.Arcade.Sdk/tools/Build.proj [hunk 3]
```diff
@@ -13,6 +13,7 @@
     DotNetBuildFromSource           Building the entire stack from source with no external dependencies.
     DotNetBuildOffline              Building without access to NuGet servers.
     DotNetOutputBlobFeedDir         Directory to publish Source Build assets to (packages, symbol pacakges, installers, etc.).
+    DotNetPublishUsingPipelines     Publish assets using Azure DevOps release pipelines.
```
Comment (chcosta) [inline src/Microsoft.DotNet.Arcade.Sdk/tools/Build.proj]: I don't think you need this

File: src/Microsoft.DotNet.Arcade.Sdk/tools/Build.proj [hunk 4]
```diff
@@ -80,6 +81,9 @@
     <_PublishToBlobStorage>false</_PublishToBlobStorage>
     <_PublishToBlobStorage Condition="'$(DotNetPublishToBlobFeed)' == 'true'">true</_PublishToBlobStorage>
 
+    <_PublishUsingPipelines>false</_PublishUsingPipelines>
+    <_PublishUsingPipelines Condition="'$(DotNetPublishUsingPipelines)' == 'true'">true</_PublishUsingPipelines>
```
Comment (chcosta) [inline src/Microsoft.DotNet.Arcade.Sdk/tools/Build.proj]: I don't think you need this

Comment (chcosta) [reply src/Microsoft.DotNet.Arcade.Sdk/tools/Build.proj]: Oh, actually, Arcade makes an explicit msbuild call to publish.proj, https://github.com/dotnet/arcade/blob/master/src/Microsoft.DotNet.Arcade.Sdk/tools/Build.proj#L270.  I forgot about that.  Sigh, in that case, I guess you need to pass the property along. :(

Comment (JohnTortugo) [reply src/Microsoft.DotNet.Arcade.Sdk/tools/Build.proj]: I don't need it. I can use `DotNetPublishUsingPipelines` to construct the properties that will get passed to Publish.proj.

File: azure-pipelines.yml [hunk 5]
```diff
@@ -47,6 +47,7 @@ jobs:
             /p:DotNetPublishBlobFeedKey=$(dotnetfeed-storage-access-key-1)
             /p:DotNetPublishBlobFeedUrl=$(_PublishBlobFeedUrl)
             /p:DotNetPublishToBlobFeed=$(_DotNetPublishToBlobFeed)
+            /p:DotNetPublishUsingPipelines=$(_PublishUsingPipelines)
```
Comment (JohnTortugo) [inline azure-pipelines.yml]: So, it looks like to make a "queue time" variable I need to create it on the visual editor. If I only reference the variable here (without creating it on the editor) I get an error. If I add the variable to the "env:" section it won't be available to changes at queue time. Reference is here: https://docs.microsoft.com/en-us/azure/devops/pipelines/process/variables?view=azure-devops&tabs=yaml%2Cbatch#allow-at-queue-time

Comment (chcosta) [reply azure-pipelines.yml]: Right, that's my understanding.  Is there a question?

> If you want this defined at queue time, You should probably just omit the "env" block from the yaml and add a variable to the build definition.

Comment (JohnTortugo) [reply azure-pipelines.yml]: Nope. Thanks

**Comment by JohnTortugo** [comment]
@chcosta / @riarenas I've validated these changes in a few builds: 
- [with PublishUsingPipelines == true](https://dnceng.visualstudio.com/internal/_build/results?buildId=117301)
- [with PublishUsingPipelines == false](https://dnceng.visualstudio.com/internal/_build/results?buildId=117398)
- [without PublishUsingPipelines variable](https://dnceng.visualstudio.com/internal/_build/results?buildId=117428)

can you please take a look again before I merge this? I'm more concerned about any possibility of breaking someone builds, improvements to the design I'll prefer to address in another issue/PR.

File: eng/common/templates/job/job.yml [hunk 6]
```diff
@@ -179,7 +179,7 @@ jobs:
       continueOnError: true
       condition: always()
     
-  - ${{ if and(eq(parameters.enablePublishBuildAssets, true), eq(parameters.runAsPublic, 'false'), ne(variables['System.TeamProject'], 'public'), notin(variables['Build.Reason'], 'PullRequest')) }}:
+  - ${{ if and(eq(parameters.enablePublishBuildAssets, true), ne(variables['_PublishUsingPipelines'], 'true'), eq(parameters.runAsPublic, 'false'), ne(variables['System.TeamProject'], 'public'), notin(variables['Build.Reason'], 'PullRequest')) }}:
```
Comment (chcosta) [inline eng/common/templates/job/job.yml]: why is this needed?

File: src/Microsoft.DotNet.Arcade.Sdk/tools/Publish.proj [hunk 7]
```diff
@@ -151,8 +149,8 @@
       ManifestBuildId="$(BUILD_BUILDNUMBER)"
       ManifestCommit="$(BUILD_SOURCEVERSION)"
       PublishFlatContainer="false"
-      AssetManifestPath="$(PipelineArtifactsAssetManifestFilePath)"
-      Condition="$(PublishToAzure) and '$(PushToAzureDevOps)' == 'true'"/>
+      AssetManifestPath="$(AssetManifestFilePath)"
+      AssetsTemporaryDirectory="$(ArtifactsDir)\..\AssetsTmpDir\$([System.Guid]::NewGuid())" />
```
Comment (chcosta) [inline src/Microsoft.DotNet.Arcade.Sdk/tools/Publish.proj]: Is there a reason this wouldn't be in `$(ArtifactsDir)tmp`?

File: eng/common/templates/job/job.yml (hunk 6 above)
Comment (JohnTortugo) [reply eng/common/templates/job/job.yml]: With the new design, Publish.proj upload the artifacts using logging commands. No need to have a separate step for it.

File: src/Microsoft.DotNet.Arcade.Sdk/tools/Publish.proj (hunk 7 above)
Comment (JohnTortugo) [reply src/Microsoft.DotNet.Arcade.Sdk/tools/Publish.proj]: I can't be under $(ArtifactsDir), the validate-sdk.ps1 script move the whole folder. It needs to be something up the tree.

File: src/Microsoft.DotNet.Build.Tasks.Feed/src/PushToAzureDevOpsArtifacts.cs [hunk 8]
```diff
@@ -77,7 +77,8 @@ public override bool Execute()
                         // removed in the future.
                         if (!Directory.Exists(AssetsTemporaryDirectory))
                         {
-                            Log.LogWarning($"Assets temporary directory {AssetsTemporaryDirectory} doesn't exist. Creating it.");
+                            Log.LogMessage(MessageImportance.High,
```
Comment (chcosta) [inline src/Microsoft.DotNet.Build.Tasks.Feed/src/PushToAzureDevOpsArtifacts.cs]: I know you don't want style feedback, but I'd say this is "MessageImportance.Low".

File: src/Microsoft.DotNet.Arcade.Sdk/tools/Publish.proj (hunk 7 above)
Comment (chcosta) [reply src/Microsoft.DotNet.Arcade.Sdk/tools/Publish.proj]: "AssetsTemporaryDirectory" should probably just be "TempDirectory" and it can default to the current directory (though we'd always want to set it to something under artifacts when we use it).  Feels weird to make this a required property.

File: eng/common/templates/job/job.yml (hunk 6 above)
Comment (chcosta) [reply eng/common/templates/job/job.yml]: That makes sense.  We'll have to try to think of some cleaner way to do this, but this is fine for now.

File: src/Microsoft.DotNet.Arcade.Sdk/tools/Publish.proj (hunk 7 above)
Comment (chcosta) [reply src/Microsoft.DotNet.Arcade.Sdk/tools/Publish.proj]: > I can't be under $(ArtifactsDir), the validate-sdk.ps1 script move the whole folder. It needs to be something up the tree.

Oh yeah, duh.  Thanks

Comment (JohnTortugo) [reply src/Microsoft.DotNet.Arcade.Sdk/tools/Publish.proj]: I totally agree that it's weird. My hope was to get rid of this thing totally. I think using TempDirectory is a better approach for now.

Comment (riarenas) [inline src/Microsoft.DotNet.Arcade.Sdk/tools/Publish.proj]: Any concerns about hitting max_path if we add that GUID to the temp dir path?

Comment (JohnTortugo) [reply src/Microsoft.DotNet.Arcade.Sdk/tools/Publish.proj]: Well, there is always that concern I guess.
